### PR TITLE
GaussianHybridFactorGraph class

### PR DESCRIPTION
**Pull request by varunagrawal**
This PR defines a new class called `GaussianHybridFactorGraph` which is the linearized version of the `HybridFactorGraph`.

As such, I moved all the elimination machinery to `GaussianHybridFactorGraph` and made other updates.

NOTE: Please review this PR carefully since I made a lot of changes to parts which were originally @dellaert's code. I tried my best to understand the overall design and what the meaning of each part is, but I could have easily made a mistake.

### CODE REVIEW

**Comment by dellaert** [comment]
So, it seems you copy/pasted?
I think the right thing to do is to make HybridFactorGraph a templated base class, so we have:

```c++
template<class FG> 
HybridFactorGraph {
// storage for DiscreteFG, DCFG, FG
};

GaussianHybridFactorGraph : public HybridFactorGraph<GaussianFactorGraph> {};
NonlinearHybridFactorGraph : public HybridFactorGraph<NonlinearFactorGraph> {};
```

**Comment by varunagrawal** [comment]
Yep pretty much copy-pasted in an effort to minimize time spent on this. That's a great suggestion, I was also thinking of the same thing but wanted to minimize changes in this PR.

**Comment by dellaert** [comment]
> Yep pretty much copy-pasted in an effort to minimize time spent on this. That's a great suggestion, I was also thinking of the same thing but wanted to minimize changes in this PR.

I think this *would* minimize PR changes, and make it easier to review. Only small changes to HFG are needed I think. And the two derived classes would be small.

**Comment by varunagrawal** [comment]
Oh I completely agree with the recommended change. I was just laying out my thought process so that we were all on the same page :)

**Comment by varunagrawal** [comment]
@dellaert if we make HybridFactorGraph a templated class, how would things like `IsNonlinear` and `IsDiscrete` work?

**Comment by varunagrawal** [comment]
Okay so I made `HybridFactorGraph` a templated base class, but I don't think it got us much. I kind of preferred the prior separation of `IsGaussian` vs `IsNonlinear` methods since they made semantic sense inside the individual subclasses.

Plus, by templating, we had to move some of the definitions to the header == larger binary size.

File: gtsam/discrete/DecisionTreeFactor.cpp
Comment (dellaert) [inline gtsam/discrete/DecisionTreeFactor.cpp:170]: Why? This should never be triggered, and we should get to root cause of the cases that *would* trigger this.

File: gtsam/hybrid/HybridFactorGraph.h
Comment (dellaert) [inline gtsam/hybrid/HybridFactorGraph.h:24]: Doxygen comment with explanation of FG

File: gtsam/hybrid/HybridFactorGraph.h
Comment (dellaert) [inline gtsam/hybrid/HybridFactorGraph.h:46]: why is that here? This class is not supposed to know about nonlinear factors

File: gtsam/hybrid/HybridFactorGraph.h
Comment (dellaert) [inline gtsam/hybrid/HybridFactorGraph.h:63]: factorGraph *of* type* FG

File: gtsam/hybrid/HybridFactorGraph.h
Comment (dellaert) [inline gtsam/hybrid/HybridFactorGraph.h:154]: Why does it not override?

File: gtsam/hybrid/HybridFactorGraph.h
Comment (dellaert) [inline gtsam/hybrid/HybridFactorGraph.h:25]: no EXPORT on templates/header only classes -> move to derived.

File: gtsam/hybrid/HybridFactorGraph.h
Comment (dellaert) [inline gtsam/hybrid/HybridFactorGraph.h:193]: Propose to remove this imperative method. Create a new instance!

File: gtsam/hybrid/HybridFactorGraph.h
Comment (dellaert) [inline gtsam/hybrid/HybridFactorGraph.h:34]: This could still contain nonlinear things of course :-( Should we template on Factor type instead ? NonlinearFactor | GaussianFactor? I'd say leave for now but something to discuss.

File: gtsam/hybrid/NonlinearHybridFactorGraph.h
Comment (dellaert) [inline gtsam/hybrid/NonlinearHybridFactorGraph.h:65]: huh ? Remove?

File: gtsam/hybrid/NonlinearHybridFactorGraph.h
Comment (dellaert) [inline gtsam/hybrid/NonlinearHybridFactorGraph.h:126]: I love this !

File: gtsam/hybrid/HybridFactorGraph.cpp
Comment (dellaert) [inline gtsam/hybrid/HybridFactorGraph.cpp:40]: template should be in header. Or just specialize it on GaussianFG and move there.

File: gtsam/hybrid/HybridFactorGraph.cpp
Comment (dellaert) [inline gtsam/hybrid/HybridFactorGraph.cpp:28]: should be in header

Comment (dellaert) [reply gtsam/hybrid/HybridFactorGraph.cpp:28]: In fact, kill this file??

File: gtsam/hybrid/GaussianHybridFactorGraph.h
Comment (dellaert) [inline gtsam/hybrid/GaussianHybridFactorGraph.h:113]: Why is this even a template? Factor is a common base class. But maybe that does not work with shared ptrs?

File: gtsam/hybrid/GaussianHybridFactorGraph.cpp
Comment (dellaert) [inline gtsam/hybrid/GaussianHybridFactorGraph.cpp:39]: Why is that here? Expected in HFG...

**Review by dellaert — COMMENTED** [review]
I *love* this. So much better!
One more iteration before approve?

File: gtsam/hybrid/HybridFactorGraph.h
Comment (varunagrawal) [reply gtsam/hybrid/HybridFactorGraph.h:46]: Okay so should I move all the `IsNonlinear` based methods to `NonlinearHybridFactorGraph` and all the `IsGaussian` based methods to `GaussianHybridFactorGraph`? Then my previous comment on the benefits of templating come up since we can't move a lot of things to the .cpp file.

File: gtsam/discrete/DecisionTreeFactor.cpp
Comment (varunagrawal) [reply gtsam/discrete/DecisionTreeFactor.cpp:170]: The lack of this causes a test to fail since we get M1 twice. There is double counting because both keys and discreteKeys in the DCFactorGraph have M1.

File: gtsam/hybrid/HybridFactorGraph.h
Comment (varunagrawal) [reply gtsam/hybrid/HybridFactorGraph.h:154]: I further override it in the subclasses but I guess it's not required.

File: gtsam/hybrid/NonlinearHybridFactorGraph.h
Comment (varunagrawal) [reply gtsam/hybrid/NonlinearHybridFactorGraph.h:65]: Ah nice catch

File: gtsam/hybrid/HybridFactorGraph.cpp
Comment (varunagrawal) [reply gtsam/hybrid/HybridFactorGraph.cpp:28]: Killed the file.

File: gtsam/hybrid/GaussianHybridFactorGraph.h
Comment (varunagrawal) [reply gtsam/hybrid/GaussianHybridFactorGraph.h:113]: We would need the template for SFINAE right? To make sure the push_back is happening to the correct factor graph?

File: gtsam/hybrid/GaussianHybridFactorGraph.cpp
Comment (varunagrawal) [reply gtsam/hybrid/GaussianHybridFactorGraph.cpp:39]: It's an override specifically for dealing with factors in `discreteGraph_` which are `DecisionTreeFactor`s.

**Comment by varunagrawal** [comment]
I'm still questioning the mileage we're getting from templating. `factorGraph_` is used in just 3 functions in `HybridFactorGraph`. All the work is still happening in the subclasses, plus we can possibly remove all the SFINAE and just specialize the `emplace_*` and `push_back*` methods to the appropriate class.
That would make the code a lot leaner IMHO.

File: gtsam/hybrid/GaussianHybridFactorGraph.cpp
Comment (dellaert) [reply gtsam/hybrid/GaussianHybridFactorGraph.cpp:41]: But that is there in HFG…

File: gtsam/hybrid/GaussianHybridFactorGraph.h
Comment (dellaert) [reply gtsam/hybrid/GaussianHybridFactorGraph.h:119]: I don’t think this method is an SFINAE. Just dynamic casts

**Review by dellaert — COMMENTED** [review]
2 things still unresolved I think.

File: gtsam/hybrid/GaussianHybridFactorGraph.cpp
Comment (varunagrawal) [reply gtsam/hybrid/GaussianHybridFactorGraph.cpp:41]: I thought there was some difference when I split the code but no test fails so yep I'll remove this. :)

File: gtsam/hybrid/GaussianHybridFactorGraph.h
Comment (varunagrawal) [reply gtsam/hybrid/GaussianHybridFactorGraph.h:119]: Let me try it out. Also FYI we have a typedef above called `sharedFactor` which I will change to `SharedFactor` so that I can use it here.

Comment (varunagrawal) [reply gtsam/hybrid/GaussianHybridFactorGraph.h:119]: Also you're correct about it not being SFINAE. I was misunderstood what was really happening here and with a fresh mind (+ without the stress of getting something running for the state estimator) this makes a lot of sense.

**Comment by varunagrawal** [comment]
I'm going to wait on #1036 to be merged before merging this so I can update this PR with all the changes and fixes from that PR.

File: gtsam/hybrid/HybridFactorGraph.h
Comment (dellaert) [inline gtsam/hybrid/HybridFactorGraph.h:18]: Why is that here?

File: gtsam/hybrid/HybridFactorGraph.h
Comment (dellaert) [reply gtsam/hybrid/HybridFactorGraph.h:42]: I like the templated version

File: gtsam/hybrid/HybridFactorGraph.h
Comment (dellaert) [reply gtsam/hybrid/HybridFactorGraph.h:193]: Does not seem to be gone

File: gtsam/hybrid/HybridFactorGraph.h
Comment (dellaert) [inline gtsam/hybrid/HybridFactorGraph.h:225]: Why is this not in GaussianHFG?

**Review by dellaert — APPROVED** [review]
I'll approve now, but "sum" still seems to be in the wrong place, I think.
I like the templated version, please keep.

File: gtsam/hybrid/HybridFactorGraph.h
Comment (varunagrawal) [reply gtsam/hybrid/HybridFactorGraph.h:225]: Ah sorry about that

File: gtsam/hybrid/HybridFactorGraph.h
Comment (varunagrawal) [reply gtsam/hybrid/HybridFactorGraph.h:193]: Oh I see what you mean. I thought create new instance meant for each member factor graph.

File: gtsam/hybrid/HybridFactorGraph.h
Comment (varunagrawal) [reply gtsam/hybrid/HybridFactorGraph.h:225]: Moved it to GaussianHFG

**Pull request merged**
